File: internal/domain/crypto/blob.go
package crypto

import (
	"errors"
	"fmt"
)

// EncryptedBlob is the versioned ciphertext envelope described in PRD §9.9.
//
// Wire format (concatenated bytes):
//
//	version (1 byte) || alg_id (1 byte) || nonce (NonceSize bytes) ||
//	ciphertext (variable) || tag (TagSize bytes)
//
// For algorithms that carry no separate tag (e.g. RSA-OAEP), Tag is nil and
// Ciphertext absorbs the tag in the underlying primitive.
type EncryptedBlob struct {
	Version    BlobVersion
	Alg        AlgID
	Nonce      []byte
	Ciphertext []byte
	Tag        []byte
}

// ErrMalformedBlob is returned when parsing a byte slice that is not a
// syntactically valid EncryptedBlob. Wraps domain.ErrInvalid via error chain
// through the caller's use case layer.
var ErrMalformedBlob = errors.New("crypto: malformed encrypted blob")

// MinBlobSize is the smallest possible blob: header + empty ciphertext for
// an algorithm with no nonce and no tag. We refuse empty trailers below this.
const MinBlobSize = 2

// Bytes returns the canonical wire-format encoding. Mutating the returned
// slice does NOT mutate the blob.
func (b EncryptedBlob) Bytes() []byte {
	out := make([]byte, 0, 2+len(b.Nonce)+len(b.Ciphertext)+len(b.Tag))
	out = append(out, byte(b.Version))
	out = append(out, byte(b.Alg))
	out = append(out, b.Nonce...)
	out = append(out, b.Ciphertext...)
	out = append(out, b.Tag...)
	return out
}

// Validate asserts the blob is well-formed for its declared algorithm.
//
// Rules:
//   - Version must equal V1.
//   - Alg must be one of the enumerated IDs.
//   - Nonce length must match the algorithm's NonceSize.
//   - Tag length must match the algorithm's TagSize.
//   - Ciphertext length is algorithm-dependent and not bounded here, but
//     must be non-empty for symmetric algorithms.
func (b EncryptedBlob) Validate() error {
	if b.Version != V1 {
		return fmt.Errorf("%w: unsupported version 0x%02x", ErrMalformedBlob, byte(b.Version))
	}
	if !b.Alg.IsValid() {
		return fmt.Errorf("%w: unknown alg 0x%02x", ErrMalformedBlob, byte(b.Alg))
	}
	if want := b.Alg.NonceSize(); len(b.Nonce) != want {
		return fmt.Errorf("%w: %s nonce len=%d want %d", ErrMalformedBlob, b.Alg, len(b.Nonce), want)
	}
	if want := b.Alg.TagSize(); len(b.Tag) != want {
		return fmt.Errorf("%w: %s tag len=%d want %d", ErrMalformedBlob, b.Alg, len(b.Tag), want)
	}
	if b.Alg == AlgAES256GCM && len(b.Ciphertext) == 0 {
		return fmt.Errorf("%w: empty ciphertext for %s", ErrMalformedBlob, b.Alg)
	}
	return nil
}

// ParseBlob decodes the wire format into an EncryptedBlob and validates it.
// Any out-of-band corruption (short buffer, unknown alg, wrong nonce length)
// results in ErrMalformedBlob.
func ParseBlob(raw []byte) (EncryptedBlob, error) {
	if len(raw) < MinBlobSize {
		return EncryptedBlob{}, fmt.Errorf("%w: input too short (%d bytes)", ErrMalformedBlob, len(raw))
	}

	b := EncryptedBlob{
		Version: BlobVersion(raw[0]),
		Alg:     AlgID(raw[1]),
	}
	if b.Version != V1 {
		return EncryptedBlob{}, fmt.Errorf("%w: unsupported version 0x%02x", ErrMalformedBlob, byte(b.Version))
	}
	if !b.Alg.IsValid() {
		return EncryptedBlob{}, fmt.Errorf("%w: unknown alg 0x%02x", ErrMalformedBlob, byte(b.Alg))
	}

	body := raw[2:]
	nonceLen := b.Alg.NonceSize()
	tagLen := b.Alg.TagSize()
	if len(body) < nonceLen+tagLen {
		return EncryptedBlob{}, fmt.Errorf("%w: body too short for %s (len=%d)", ErrMalformedBlob, b.Alg, len(body))
	}

	// Layout: nonce || ciphertext || tag (tag is the suffix).
	if nonceLen > 0 {
		b.Nonce = append([]byte(nil), body[:nonceLen]...)
		body = body[nonceLen:]
	}
	ctLen := len(body) - tagLen
	b.Ciphertext = append([]byte(nil), body[:ctLen]...)
	if tagLen > 0 {
		b.Tag = append([]byte(nil), body[ctLen:]...)
	}

	if err := b.Validate(); err != nil {
		return EncryptedBlob{}, err
	}
	return b, nil
}

File: web/src/shared/crypto/blob.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

/**
 * EncryptedBlob — versioned ciphertext envelope (PRD §9.9, C5).
 *
 * Wire format (concatenated bytes):
 *   version (1B) || alg_id (1B) || nonce (NonceSize) || ciphertext (var) || tag (TagSize)
 *
 * Byte-identical to Go's domain/crypto.EncryptedBlob.
 */

import {
  BLOB_VERSION,
  type AlgID,
  isValidAlgId,
  nonceSize,
  tagSize,
} from "./algorithm.js";

export class MalformedBlobError extends Error {
  constructor(message: string) {
    super(`crypto: malformed encrypted blob: ${message}`);
    this.name = "MalformedBlobError";
  }
}

export interface EncryptedBlob {
  version: number;
  alg: AlgID;
  nonce: Uint8Array;
  ciphertext: Uint8Array;
  tag: Uint8Array;
}

/** Validate an EncryptedBlob against its algorithm's invariants. */
export function validateBlob(b: EncryptedBlob): void {
  if (b.version !== BLOB_VERSION) {
    throw new MalformedBlobError(
      `unsupported version 0x${b.version.toString(16).padStart(2, "0")}`,
    );
  }
  if (!isValidAlgId(b.alg as number)) {
    throw new MalformedBlobError(
      `unknown alg 0x${(b.alg as number).toString(16).padStart(2, "0")}`,
    );
  }

  const wantNonce = nonceSize(b.alg);
  if (b.nonce.length !== wantNonce) {
    throw new MalformedBlobError(
      `nonce len=${b.nonce.length} want ${wantNonce}`,
    );
  }

  const wantTag = tagSize(b.alg);
  if (b.tag.length !== wantTag) {
    throw new MalformedBlobError(`tag len=${b.tag.length} want ${wantTag}`);
  }

  if (b.alg === 0x01 && b.ciphertext.length === 0) {
    throw new MalformedBlobError("empty ciphertext for AES-256-GCM");
  }
}

/** Serialize an EncryptedBlob to wire format. */
export function serializeBlob(b: EncryptedBlob): Uint8Array {
  validateBlob(b);
  const out = new Uint8Array(
    2 + b.nonce.length + b.ciphertext.length + b.tag.length,
  );
  out[0] = b.version;
  out[1] = b.alg;
  let offset = 2;
  out.set(b.nonce, offset);
  offset += b.nonce.length;
  out.set(b.ciphertext, offset);
  offset += b.ciphertext.length;
  out.set(b.tag, offset);
  return out;
}

/** Parse wire-format bytes into an EncryptedBlob. */
export function parseBlob(raw: Uint8Array): EncryptedBlob {
  if (raw.length < 2) {
    throw new MalformedBlobError(`input too short (${raw.length} bytes)`);
  }

  const version = raw[0]!;
  const algByte = raw[1]!;

  if (version !== BLOB_VERSION) {
    throw new MalformedBlobError(
      `unsupported version 0x${version.toString(16).padStart(2, "0")}`,
    );
  }
  if (!isValidAlgId(algByte)) {
    throw new MalformedBlobError(
      `unknown alg 0x${algByte.toString(16).padStart(2, "0")}`,
    );
  }

  const alg = algByte as AlgID;
  const nonceLen = nonceSize(alg);
  const tagLen = tagSize(alg);
  const body = raw.subarray(2);

  if (body.length < nonceLen + tagLen) {
    throw new MalformedBlobError(
      `body too short for alg (len=${body.length})`,
    );
  }

  const nonce = body.slice(0, nonceLen);
  const remaining = body.subarray(nonceLen);
  const ctLen = remaining.length - tagLen;
  const ciphertext = remaining.slice(0, ctLen);
  const tag = tagLen > 0 ? remaining.slice(ctLen) : new Uint8Array(0);

  const blob: EncryptedBlob = { version, alg, nonce, ciphertext, tag };
  validateBlob(blob);
  return blob;
}
